Feat/cpu friendly attn implementation

BouajilaHamza opened the pull request:
**Fixes:** N/A (enables CPU‐only execution)

### Changes proposed in this pull request

* **Policy-level API**

  * Added an `attn_implementation` parameter to `Gr00tPolicy.__init__` (options: `"auto"`, `"eager"`, `"flash_attention_2"`).
  * In `_load_model()`, automatically select `None` (default) on GPU and `"eager"` on CPU when `attn_implementation="auto"`, or respect explicit overrides.

* **Core model loader**

  * Updated `GR00T_N1.from_pretrained()` to pop and forward `attn_implementation` exactly once, and inject it into the HF `AutoConfig` before calling `super()`.

* **Backbone integration**

  * Added support for in‑memory attn_implementation propagation in EagleBackbone—now passes the user’s preference directly into both AutoConfig.from_pretrained() and AutoModel.from_config() instead of requiring any manual JSON edits.

* **Nested config propagation**

  * In `Eagle2ChatConfig`, propagate `attn_implementation` from `kwargs` into both `vision_config` and `llm_config` before instantiating `SiglipVisionConfig` and `LlamaConfig`.

* **Tests**

  * Added/updated unit tests for `Gr00tPolicy` to verify `"auto"`, `"eager"`, and `"flash_attention_2"` behaviors under mocked GPU/CPU conditions, patching out file I/O.
  * Added unit tests for `EagleBackbone` to confirm `attn_implementation` is passed into HF APIs.
  * Added unit tests for `Eagle2ChatConfig` to confirm nested config fields are set when provided.

---

## Before submitting

* [x] I've read and followed all steps in the [[Making a pull request](https://github.com/NVIDIA/Isaac-GR00T/blob/main/CONTRIBUTING.md#making-a-pull-request)](https://github.com/NVIDIA/Isaac-GR00T/blob/main/CONTRIBUTING.md#making-a-pull-request) section of the CONTRIBUTING docs.
* [x] I've updated or added all relevant docstrings (in `policy.py`, `eagle_backbone.py`, and the configuration classes).
* [x] If this PR fixes a bug, I've added tests that would fail without my fix (simulating FlashAttention error on CPU).
* [x] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality (CPU/GPU auto selection and explicit overrides).

---

**Notes for reviewers:**

* On CPU-only machines, `attn_implementation="auto"` now correctly falls back to `"eager"`, preventing import errors from `flash_attn`.
* On GPU, the default behavior (`None`) allows the library to choose the fastest available backend (FlashAttention2 or SDPA).
* No on-disk modifications are done; all behavior is in-memory and consistent with HF best practices.

BouajilaHamza committed:
implement cpu friendly attention implementation auto detection of gpu/cpu

BouajilaHamza committed:
update detecting hardware and choosing attn_implementation accordingly

BouajilaHamza committed:
update detecting hardware and choosing attn_implementation accordingly

BouajilaHamza committed:
add test_config_attn_propagation

BouajilaHamza committed:
add test_config_attn_propagation

BouajilaHamza committed:
add testing the backbone attn implementation propagation

BouajilaHamza committed:
add testing the backbone attn implementation propagation

BouajilaHamza committed:
add unit test for testing policy with flash_attention_2 on cuda

BouajilaHamza committed:
add unit test for testing policy with flash_attention_2 on cuda

BouajilaHamza committed:
add other test cases for policy auto+cuda / auto+cpu / explicit eager+cpu

BouajilaHamza committed:
add other test cases for policy auto+cuda / auto+cpu / explicit eager+cpu

BouajilaHamza committed:
remove useless comments

BouajilaHamza committed:
remove useless comments

BouajilaHamza committed:
fixing linting , sorting and ruff checks

BouajilaHamza committed:
fixing linting , sorting and ruff checks

BouajilaHamza commented:
cc @youliangtan  for review
thanks

youliangtan commented on gr00t/model/policy.py:
```diff
@@ -231,10 +235,19 @@ def _check_state_is_batched(self, obs: Dict[str, Any]) -> bool:
-        model.to(device=self.device)  # type: ignore
-
+        # decide actual setting
+        if self.attn_implementation_preference == "auto":
```
Can remove this if else statement here, and just simply parse in the `attn_implementaion` into it

youliangtan commented on gr00t/model/policy.py:
```diff
@@ -231,10 +235,19 @@ def _check_state_is_batched(self, obs: Dict[str, Any]) -> bool:
+        if self.attn_implementation_preference == "auto":
+            chosen = None if torch.cuda.is_available() else "eager"
+        else:
+            chosen = self.attn_implementation_preference
```
Can remove this if else statement block here, and just simply parse in the `attn_implementation_preference` into `from_pretrained`

youliangtan commented on uv.lock:
Remove this file, since not needed?

youliangtan commented on getting_started/1_gr00t_inference.ipynb:
stash the commit here if diff are not needed

youliangtan reviewed (changes requested):
Thanks for the PR. Briefly checking this, this presume this would also be helpful if flash-attn is not installed too? 

Overall looks good to me. I dropped in some minor change requests too. Make sure the lining works `isort . &&  black .`

BouajilaHamza committed:
remove if else statement and parse the attn_imeplementation directly

BouajilaHamza committed:
remove uv.lock

BouajilaHamza committed:
remove uv.lock

BouajilaHamza committed:
Merge branch 'feat/cpu-friendly-attn-implementation' of https://github.com/BouajilaHamza/Isaac-GR00T into feat/cpu-friendly-attn-implementation

youliangtan commented:
Getting some styling error in the CI, do refer to https://github.com/NVIDIA/Isaac-GR00T/blob/main/CONTRIBUTING.md  . Thanks!

BouajilaHamza commented:
@youliangtan yes i have noticed that i will check soon

BouajilaHamza committed:
format using black

BouajilaHamza committed:
sort code with isort

BouajilaHamza committed:
remove duplicated code in configuration_eagle_chat.py

BouajilaHamza committed:
Revert "update detecting hardware and choosing attn_implementation accordingly"

This reverts commit dbfc80465a9b03ee051a7e4665af276093408283.

BouajilaHamza committed:
add attn_implementation in configuration file after reverting

BouajilaHamza committed:
update testing logic of auto set attn_implementation inside the groot_n1.from_trained method and inserted correctly in model configuration

BouajilaHamza committed:
Revert "update detecting hardware and choosing attn_implementation accordingly"

This reverts commit dbfc80465a9b03ee051a7e4665af276093408283.

BouajilaHamza committed:
format and sort with black and isort

BouajilaHamza committed:
fix linting

youliangtan commented:
I breifly tested it by loading a trained lora checkpoint, but getting this error. The previous main works well.
```
Flash Attention 2 available and selected.
Using attention implementation: flash_attention_2
Traceback (most recent call last):
  File "/home/youliang/gear/clean_repo/Isaac-GR00T/scripts/inference_service.py", line 73, in <module>
    policy = Gr00tPolicy(
  File "/home/youliang/gear/clean_repo/Isaac-GR00T/gr00t/model/policy.py", line 112, in __init__
    self._load_model(model_path)
  File "/home/youliang/gear/clean_repo/Isaac-GR00T/gr00t/model/policy.py", line 239, in _load_model
    model = GR00T_N1.from_pretrained(
  File "/home/youliang/gear/clean_repo/Isaac-GR00T/gr00t/model/gr00t_n1.py", line 248, in from_pretrained
    config = AutoConfig.from_pretrained(pretrained_model_name_or_path, trust_remote_code=True)
  File "/home/youliang/miniconda3/envs/groot-release/lib/python3.10/site-packages/transformers/models/auto/configuration_auto.py", line 1038, in from_pretrained
    raise ValueError(
ValueError: Unrecognized model in /home/youliang/checkpoints/my-lora-checkpoint/checkpoint-9500/. Should have a `model_type` key in its config.json, or contain one of the following strings in its name: albert, align, altclip, audio-spectrogram-transformer, autoformer, bark, bart, beit, bert, bert-generation, big_bird, bigbird_pegasus, biogpt, bit, blenderbot, blenderbot-small, blip, blip-2, bloom, bridgetower, bros, camembert, canine, chameleon, chinese_clip, chinese_clip_vision_model, clap, clip, clip_text_model, clip_vision_model, clipseg, clvp, code_llama, codegen, cohere, conditional_detr, convbert, convnext, ...
tapas, time_series_transformer, timesformer, timm_backbone, trajectory_transformer, transfo-xl, trocr, tvlt, tvp, udop, umt5, unispeech, unispeech-sat, univnet, upernet, van, video_llava, videomae, vilt, vipllava, vision-encoder-decoder, vision-text-dual-encoder, visual_bert, vit, vit_hybrid, vit_mae, vit_msn, vitdet, vitmatte, vits, vivit, wav2vec2, wav2vec2-bert, wav2vec2-conformer, wavlm, whisper, xclip, xglm, xlm, xlm-prophetnet, xlm-roberta, xlm-roberta-xl, xlnet, xmod, yolos, yoso, zoedepth, gr00t_n1
```

Base pretrained model, and non-lora checkpoint loading works. any idea for a quick fix?

not-lain commented:
Hi @youliangtan 
the issue stems from your `config.json` file, by default all config files should have a `model_type` or a `auto_map` parameter (or both which should reroute to the auto_map ) which is used to identify which config class to use with the AutoConfig, you can take a look at my config.json for a custom model which you can use as a reference at https://huggingface.co/not-lain/MyRepo/blob/main/config.json 

feel free to share your `config.json` with me or for better debugging on my end a HF repo and i'll try to help you fix this issue.

(This issue is not related to this pr and only related to the config file) 

it also might be because your loading script of the peft model is either wrong (loading the model config on the peft config file) or something went sideways along the way.

youliangtan commented:
@not-lain  This is the lora checkpoint `adapter_config.json` which works on previously. Just want to make sure this still works before merging in to prevent regression
```
{
  "alpha_pattern": {},
  "auto_mapping": null,
  "base_model_name_or_path": "/home/youliang/.cache/huggingface/hub/models--nvidia--GR00T-N1-2B/snapshots/32e1fd2507f7739fad443e6b449c8188e0e02fcb",
  "bias": "none",
  "eva_config": null,
  "exclude_modules": null,
  "fan_in_fan_out": false,
  "inference_mode": true,
  "init_lora_weights": true,
  "layer_replication": null,
  "layers_pattern": null,
  "layers_to_transform": null,
  "loftq_config": {},
  "lora_alpha": 128,
  "lora_bias": false,
  "lora_dropout": 0.1,
  "megatron_config": null,
  "megatron_core": "megatron.core",
  "modules_to_save": null,
  "peft_type": "LORA",
  "r": 64,
  "rank_pattern": {},
  "revision": null,
  "target_modules": [
    "to_v",
    "to_k",
    "to_q",
    "k_proj",
    "q_proj",
    "v_proj"
  ],
  "task_type": "CAUSAL_LM",
  "use_dora": false,
  "use_rslora": false
}
```

not-lain commented:
@youliangtan 
could you also share your `config.json` for the model as well and any other reproducibles?

generally to load the peft adapter even for a custom model you run something like this: 
```python
from peft import PeftModel

base_model = ModelClass.from_pretrained("model_dir") # or base_model  = ModelClass() or any other initialization script that you have 
model = PeftModel.from_pretrained(base_model, "adapter_dir")
```
if the model has been loaded correctly, there is no issue on the PEFT side

youliangtan commented:
@not-lain the base model is the same as the one in https://huggingface.co/nvidia/GR00T-N1-2B

not-lain commented:
thanks, will run this pr on my end and check if there are any issues and get back to you

BouajilaHamza commented:
@youliangtan  Okey i will check it soon 
Insha allah

BouajilaHamza commented:
@youliangtan  may i ask how you loaded the model to reproduce the issue :
i used this and did not get the error :
from peft import PeftModel

base_model = ModelClass.from_pretrained("model_dir") # or base_model  = ModelClass() or any other initialization script that you have 
model = PeftModel.from_pretrained(base_model, "adapter_dir")

BouajilaHamza commented:
@youliangtan  i have tried to reproduce the error i made many steps and here is what i found : 
i am using sdpa so i can run gr00t on kaggle , updated the config.json manually for now

1. i loaded gr00t using **peft +sdpa** on kaggle it worked **fine** **main** branch
2. i loaded gr00t using **policy+sdpa** on kaggle got same **error** in **main** branch
3. i loaded gr00t using **peft+sdpa** on kaggle worked **fine** on **BouajilaHamza:feat/cpu-friendly-attn-implementation** 
4. i loaded gr00t using **policy+sdpa** on kaggle got same error on **BouajilaHamza:feat/cpu-friendly-attn-implementation**

BouajilaHamza committed:
update Gr00t model from_pretrained config implementation , remove forcing custom config and implement Patch In-Place After Loading the model add sdpa implementation

BouajilaHamza committed:
move selecting attn_implementation logic  under utils, add checking flash_ttn_2 chardware comptability , else falling back sdpa  in the logic

BouajilaHamza committed:
update tests , add loading lora checkpoint test

BouajilaHamza commented:
@youliangtan 
I have investigated more the issue .
I was enforcing custom config into GR00T_N1 model which bypassed the internal behaviour of checking lora checkpoints in from_pretrained method .
Instead i updated to attn implementation after loading the model just updating the attn implmentation to the loading the config instead of creating another .
Also i have updated the tests and added sdpa because there is no need to create another PR it is the same logic .
Thank you!

BouajilaHamza committed:
remove local used dependencies

BouajilaHamza committed:
Merge branch 'main' of https://github.com/NVIDIA/Isaac-GR00T into feat/cpu-friendly-attn-implementation
